src/compare.rs

```diff
@@ -34,20 +34,34 @@ fn cmp_segment(lhs: &str, rhs: &str, mode: UnderscoreOrder) -> Ordering {
         (false, false) => {}
     }
 
-    if mode == UnderscoreOrder::Last {
-        match count_leading_underscores(lhs).cmp(&count_leading_underscores(rhs)) {
-            Ordering::Equal => {}
-            non_eq => return non_eq,
-        }
-    }
+    let mut lhs_atoms = iter_atoms(&lhs);
+    let mut rhs_atoms = iter_atoms(&rhs);
 
-    let lhs = lhs.to_ascii_lowercase();
-    let rhs = rhs.to_ascii_lowercase();
+    loop {
+        let left = lhs_atoms.next();
+        let right = rhs_atoms.next();
 
-    // For now: asciibetical ordering.
-    lhs.cmp(&rhs)
-}
+        let (left, right) = match (left, right) {
+            (None, None) => return Ordering::Equal,
+            (None, Some(_)) => return Ordering::Greater,
+            (Some(_), None) => return Ordering::Less,
+            (Some(left), Some(right)) => (left, right),
+        };
 
-fn count_leading_underscores(segment: &str) -> usize {
-    segment.chars().take_while(|c| *c == '_').count()
+        // Compare underscores with respect to mode.
+        match (&left, &right) {
+            (Atom::Underscore(l), Atom::Underscore(r)) => match l.cmp(&r) {
+                Ordering::Equal => continue,
+                non_eq => return non_eq,
+            },
+            (Atom::Underscore(_), _) if mode == UnderscoreOrder::First => return Ordering::Greater,
+            (Atom::Underscore(_), _) => return Ordering::Less,
+            (_, Atom::Underscore(_)) if mode == UnderscoreOrder::First => return Ordering::Less,
+            (_, Atom::Underscore(_)) => return Ordering::Greater,
```

Comment:
UnderscoreOrder should only be applied to leading underscores. Things like GNU sort and vim's sort like to put underscores first (`under__score < under_score`) but in Rust we often want leading underscores (`__Nonexhaustive`) last as a special case.